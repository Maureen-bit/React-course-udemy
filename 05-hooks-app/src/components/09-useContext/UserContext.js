import { createContext } from 'react'

/* un context es un order higher component ya que va a contener m√°s componentes adentro */
export const UserContext = createContext(null);
