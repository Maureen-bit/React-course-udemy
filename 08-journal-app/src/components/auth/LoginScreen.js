import React, { useMemo } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { startAuthenticationWithEmailAndPassword, startGoogleSignIn } from './thunks';

export const LoginScreen = () => {

  const { status, errorMessage } = useSelector( state => state.auth );

  const dispatch = useDispatch();
  const { email, password, onInputChange } = useForm({
    email: 'maureen@gmail.com',
    password: 123456
  });

  const isAuthenticating = useMemo(() => status === 'checking', [status])

  const onSubmit = (event) => {
    event.preventDefault();
    console.log({email, password})
    dispatch( startAuthenticationWithEmailAndPassword(email, password) );
  }

  const onGoogleSignIn = () => {
    console.log('ON GOOGLE SIGN IN')
    dispatch( startGoogleSignIn() );
  }


  return (
    <>
      <h3 className='auth__title'>Login</h3>
      <form onSubmit={onSubmit}>
        <input 
          type='text'
          placeholder='email'
          name='email'
          className='auth__input'
          autoComplete='off'
          onChange={onInputChange}
        />
        <input 
          type='password'
          placeholder='password'
          name='password'
          className='auth__input'
          onChange={onInputChange}
        />

        <button 
         type='submit'
         className='btn btn-primary btn-block'
         disabled={isAuthenticating}
        >
          Login
        </button>
        {errorMessage && errorMessage}
        <div className='auth__social-networks'>
          <p>Login with social network</p>
          <button 
          className='btn btn-primary btn-block'
          onClick={onGoogleSignIn}
          disabled={isAuthenticating}
          >
            Google
          </button>
        </div>

        <Link 
          to='/auth/register'
          className='link'
        >  
        Create new account
        </Link>
      </form>
    </>
  )
}
