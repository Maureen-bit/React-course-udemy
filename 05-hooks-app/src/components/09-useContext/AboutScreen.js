import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const AboutScreen = () => {
  const { user, setUser } = useContext(UserContext);
  const handleClick = () => {
    /* Cuando el context sufre una modificaci√≥n notifica a todos los hijos sobre el cambio y hace que se 
    redibuje cada una de las partes afectadas */
    setUser({})
  }
  return (
    <div>
        <h1>AboutScreen</h1>
        <hr/>
        <pre>
            {JSON.stringify(user, null, 3)}
        </pre>
        <button className='btn btn-warning' 
        onClick={handleClick}
        > 
          Logout 
        </button>
    </div>
  )
}
