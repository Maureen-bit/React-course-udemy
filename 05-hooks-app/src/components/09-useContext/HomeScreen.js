import React, {useContext} from 'react'
import { UserContext } from './UserContext'

export const HomeScreen = () => {
  /* Obtener el valor de nuestro useContext creado para guardar datos del usuario */
  const { user } = useContext(UserContext);
  return (
    <div>
        <h1>HomeScreen</h1>
        <hr/>
        <pre>
            {JSON.stringify(user, null, 3)}
        </pre>
    </div>
  )
}
