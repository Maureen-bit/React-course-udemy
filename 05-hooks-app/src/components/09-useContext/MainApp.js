import React, { useState } from 'react';
import { AppRouter } from './AppRouter';
import { UserContext } from './UserContext';

export const MainApp = () => {
    /* UserContext: va a envolver todos los componentes que encuentre en AppRouter, es decir que, la información
    que se guarde en este context estará disponible para todos estos hijos.
    Lo que se encuentra en la propiedad value es lo que voy a compartir  */
    const [user, setUser] = useState({})

  return (
    <UserContext.Provider value={{user, setUser}}>
        <AppRouter />
    </UserContext.Provider>
  )
}
