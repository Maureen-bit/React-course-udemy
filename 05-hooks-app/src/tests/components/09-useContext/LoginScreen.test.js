import React from 'react';
import { mount } from 'enzyme';
import { LoginScreen } from '../../../components/09-useContext/LoginScreen';
import { UserContext } from '../../../components/09-useContext/UserContext';

describe('Tests in <LoginScreen /> Component', () => { 
    const setUser = jest.fn();
    const param = {
        id: 123456, 
        name:"Maureen y Candy", 
        email: "maureen@gmail.com" 
    };
    const wrapper = mount( 
        <UserContext.Provider value={{
            setUser: setUser
        }}>
            <LoginScreen />
        </UserContext.Provider> 
    );
    test('It should renders correctly', () => { 
        expect(wrapper).toMatchSnapshot();
     });

     test('it should call setUser with an expected argument', () => { 
        /* paréntesis al final significa ejecución de la función que se encuentra en la propiedad del onClick */
        wrapper.find('button').prop('onClick')();
        expect(setUser).toHaveBeenCalledWith(param);
      })
 })