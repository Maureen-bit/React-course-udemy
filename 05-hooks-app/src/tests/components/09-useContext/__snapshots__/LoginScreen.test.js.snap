// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <LoginScreen /> Component It should renders correctly 1`] = `
<div>
  <h1>
    LoginScreen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
